src/replication.c

```diff
@@ -1576,13 +1588,47 @@ pause_and_timed_wait_for_ongoing_ios(spec_t *spec, int sec)
 	return ret;
 }
 
+#define BUILD_VOL_CMD_DATA(_spec, _opcode, _snapname, _size, _data, _data_len)				\
+	do {												\
+		zvol_op_resize_data_t *resize_cmd;							\
+		switch (_opcode) {	 								\
+			case ZVOL_OPCODE_SNAP_CREATE:							\
+			case ZVOL_OPCODE_SNAP_DESTROY:							\
+				_data_len = strlen(_spec->volname) + strlen(_snapname) + 2;		\
+				_data = (char *)malloc(_data_len);					\
+				snprintf(_data, _data_len, "%s@%s", _spec->volname, _snapname);		\
+				break;									\
+			case ZVOL_OPCODE_RESIZE:							\
+				_data_len = sizeof(zvol_op_resize_data_t);				\
+				resize_cmd = malloc(_data_len);						\
+				memset(resize_cmd, 0, _data_len);					\
+				strncpy(resize_cmd->volname, _spec->volname, MAX_NAME_LEN);		\
+				resize_cmd->size = _size;						\
+				_data = resize_cmd;							\
+				break;									\
+			default:									\
+				MTX_UNLOCK(&spec->rq_mtx);						\
+				REPLICA_LOG("Please handle new volume operation opcode: %d\n", opcode);	\
+				return false;								\
+			}										\
+	} while (0)											\
+
 int istgt_lu_destroy_snapshot(spec_t *spec, char *snapname)
 {
 	replica_t *replica;
-	TAILQ_FOREACH(replica, &spec->rq, r_next)
-		send_replica_snapshot(spec, replica, 0, snapname, ZVOL_OPCODE_SNAP_DESTROY, NULL);
-	TAILQ_FOREACH(replica, &spec->non_quorum_rq, r_non_quorum_next)
-		send_replica_snapshot(spec, replica, 0, snapname, ZVOL_OPCODE_SNAP_DESTROY, NULL);
+	zvol_op_code_t opcode = ZVOL_OPCODE_SNAP_DESTROY;
+	uint64_t size = 0;
+	size_t data_len = 0;
+	void *data = NULL;
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		BUILD_VOL_CMD_DATA(spec, opcode, snapname, size, data, data_len);
```

Comment:
@mittachaitu We are not taking a lock in this fn. Let's keep, release and return, outside of MACRO only. With that, it will also increase the readability.
Please let me know you suggestion.

Reply:
updated the MACRO